### ISSUE TITLE
Possible issues with Spring's AOP class proxy with CGLIB

### ISSUE THREAD

**Issue by lombokissues** [rspilker/trackertest#67]
*Migrated from Google Code (issue 28)*

**Comment by lombokissues** [comment]
:bust_in_silhouette: **warchildwts** &nbsp; :clock8: Aug 13, 2009 at 04:54 UTC

<b>What steps will reproduce the problem?</b>
1.  Create a class that uses lombok's @Data annotation
2.  Create a spring aop proxy around that class that calls it's toString
method (you may or may not need to set proxy-target-class=&quot;true&quot;).
3.  Write a test that causes one of the generated (toString or equals)
methods to fire.

<b>What is the expected output? What do you see instead?</b>
I would expect to see the lombok toString instead of object.toString, but
I'm seeing the object.toString

<b>What version of the product are you using? On what operating system?</b>
eclipse 3.5
linux x86_64
Java 1.6
CGLIB 2.3
Lombok 0.8.2

<b>Please provide any additional information below.</b>
I guess I could see how this could happen though.  Since CGLIB likely isn't
going to be creating classes with the lombok code in them.  If I'm
understanding this right, I don't know if it is something that can be
solved since the proxy classes generated by spring's AOP use cglib, not javac.

**Comment by lombokissues** [comment]
:bust_in_silhouette: **warchildwts** &nbsp; :clock8: Aug 13, 2009 at 07:16 UTC

I'm working on trying to get a minimal project that can be used to repro this.  I'll
attach it here when I have one.

**Comment by lombokissues** [comment]
:bust_in_silhouette: **warchildwts** &nbsp; :clock8: Aug 13, 2009 at 16:32 UTC

Ok, after looking at this a little bit more, I don't think it's what I was initially
thinking.  A sample project that I created worked correctly.  It's looking like the
proxy generated by CGLIB is not getting advised by the AOP class for calls to equals
and hashCode methods.

Still trying to track this one down.

**Comment by lombokissues** [comment]
:bust_in_silhouette: **warchildwts** &nbsp; :clock8: Aug 16, 2009 at 18:20 UTC

Ok, I've finally got a repo project.  I'm attaching it here.  From what I understand,
the hashCode, equals and toString methods should be implemented by lombok in the
.class file generated by the compiler.  So, at runtime when spring tries to use CGLIB
to create the proxy of the class, it should get lombok's implementation of the
equals, hashCode and toString, but it doesn't seem to.  It looks like it's getting
the one from object instead.  I've tried this with and without the overrides in the
interface (was just a test to try to narrow it down further).  I've attached both tgz
& zipped versions of the same project.  

This project requires the m2eclipse plugin to be installed.  I also have the spring
tools installed, but I don't know that those are required to work with the project.
